third/nmh/man/prompter.man
.\"
.\" %nmhwarning%
.\" $Id: prompter.man,v 1.1.1.1 1999-02-07 18:14:22 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH PROMPTER %manext1% MH.6.8 [%nmhversion%]
.SH NAME
prompter \- prompting editor front-end for nmh
.SH SYNOPSIS
.in +.5i
.ti -.5i
prompter
\%[\-erase\ chr]
\%[\-kill\ chr]
\%[\-prepend] \%[\-noprepend]
\%[\-rapid] \%[\-norapid]
\%[\-doteof] \%[\-nodoteof]
file
\%[\-version]
\%[\-help]
.in -.5i
.SH DESCRIPTION
\fIPrompter\fR is an editor front\-end for \fInmh\fR which allows rapid
composition of messages.  This program is not normally invoked directly by
users but takes the place of an editor and acts as an editor front\-end.
It operates on an RFC\-822 style message draft skeleton specified by
file, normally provided by the nmh commands \fIcomp\fR, \fIdist\fR,
\fIforw\fR, or \fIrepl\fR.

\fIPrompter\fR is particularly useful when composing messages over slow
network or modem lines.  It is an \fInmh\fR program in that it can have
its own profile entry with switches, but it is not invoked directly by
the user.  The commands \fIcomp\fR, \fIdist\fR, \fIforw\fR, and \fIrepl\fR
invoke \fIprompter\fR as an editor, either when invoked with
`\-editor\ prompter', or by the profile entry \*(lqEditor:\ prompter\*(rq,
or when given the command `edit\ prompter' at the \*(lqWhat now?\*(rq prompt.

For each empty component \fIprompter\fR finds in the draft, the user
is prompted for a response; A <RETURN> will cause the whole component
to be left out.  Otherwise, a `\\' preceding a <RETURN> will continue
the response on the next line, allowing for multiline components.
Continuation lines \fBmust\fR begin with a space or tab.

Each non\-empty component is copied to the draft and displayed on the
terminal.

The start of the message body is denoted by a blank line or a line
of dashes.  If the body is non\-empty, the prompt, which isn't written
to the file, is

    \*(lq--------Enter additional text\*(rq,

or (if `\-prepend' was given)

    \*(lq--------Enter initial text\*(rq.

Message\-body typing is terminated with an end\-of\-file (usually
CTRL\-D).  With the `\-doteof' switch, a period on a line all by itself
also signifies end\-of\-file.  At this point control is returned to
the calling program, where the user is asked \*(lqWhat now?\*(rq.
See \fIwhatnow\fR for the valid options to this query.

By using the `\-prepend' switch, the user can add type\-in to the
beginning of the message body and have the rest of the body follow.
This is useful for the \fIforw\fR command.

By using the `\-rapid' switch, if the draft already contains text in
the message\-body, it is not displayed on the user's terminal.  This is
useful for low\-speed terminals.

The line editing characters for kill and erase may be specified by the
user via the arguments `\-kill\ chr' and `\-erase\ chr', where chr may
be a character; or `\\nnn', where \*(lqnnn\*(rq is the octal value for
the character.

An interrupt (usually CTRL\-C) during component typing will abort
\fIprompter\fR and the \fInmh\fR command that invoked it.  An interrupt
during message\-body typing is equivalent to CTRL\-D, for historical
reasons.  This means that \fIprompter\fR should finish up and exit.

The first non\-flag argument to \fIprompter\fR is taken as the name of
the draft file, and subsequent non\-flag arguments are ignored.
.\" (\fIRepl\fR invokes editors with two file arguments:
.\" the draft file name and the replied\-to message file name.)
.Fi
^$HOME/\&.mh\(ruprofile~^The user profile
^/tmp/prompter*~^Temporary copy of message
.Pr
prompter\-next:	To name the editor to be used on exit from \fIprompter\fR
.Ps
^Msg\-Protect:~^To set mode when creating a new draft
.Sa
comp(1), dist(1), forw(1), repl(1), whatnow(1)
.De
`\-prepend'
.Ds
`\-norapid'
.Ds
`\-nodoteof'
.Co
None
.Hh
The `\-rapid' option is particularly useful with \fIforw\fP, and
`\-noprepend' is useful with \fIcomp\ \-use\fP.

The user may wish to link \fIprompter\fR under several names (e.g.,
\*(lqrapid\*(rq) and give appropriate switches in the profile entries
under these names (e.g., \*(lqrapid: -rapid\*(rq).  This facilitates
invoking prompter differently for different \fInmh\fP commands (e.g.,
\*(lqforw: -editor rapid\*(rq).
.Bu
\fIPrompter\fR uses \fIstdio\fR\0(3), so it will lose if you edit files
with nulls in them.
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/prompter.c

/*
 * prompter.c -- simple prompting editor front-end
 *
 * $Id: prompter.c,v 1.1.1.1 1999-02-07 18:14:16 danw Exp $
 */

#include <h/mh.h>
#include <fcntl.h>
#include <h/signals.h>
#include <errno.h>
#include <signal.h>
#include <setjmp.h>

#ifdef HAVE_TERMIOS_H
# include <termios.h>
#else
# ifdef HAVE_TERMIO_H
#  include <termio.h>
# else
#  include <sgtty.h>
# endif
#endif

#define	QUOTE '\\'

#ifndef	CKILL
# define CKILL '@'
#endif

#ifndef	CERASE
# define CERASE '#'
#endif

static struct swit switches[] = {
#define	ERASESW	0
    { "erase chr", 0 },
#define	KILLSW	1
    { "kill chr", 0 },
#define	PREPSW	2
    { "prepend", 0 },
#define	NPREPSW	3
    { "noprepend", 0 },	
#define	RAPDSW	4
    { "rapid", 0 },
#define	NRAPDSW	5
    { "norapid", 0 },
#define	BODYSW	6
    { "body", -4 },
#define	NBODYSW	7
    { "nobody", -6 },
#define	DOTSW	8
    { "doteof", 0 },
#define	NDOTSW	9
    { "nodoteof", 0 },
#define VERSIONSW 10
    { "version", 0 },
#define	HELPSW	11
    { "help", 4 },
    { NULL, 0 }
};

extern int errno;

#ifdef HAVE_TERMIOS_H
static struct termios tio;
# define ERASE tio.c_cc[VERASE]
# define KILL  tio.c_cc[VKILL]
# define INTR  tio.c_cc[VINTR]
#else
# ifdef HAVE_TERMIO_H
static struct termio tio;
#  define ERASE tio.c_cc[VERASE]
#  define KILL  tio.c_cc[VKILL]
#  define INTR  tio.c_cc[VINTR]
# else
static struct sgttyb tio;
static struct tchars tc;
#  define ERASE tio.sg_erase
#  define KILL  tio.sg_kill
#  define INTR  tc.t_intrc
# endif
#endif

static int wtuser = 0;
static int sigint = 0;
static jmp_buf sigenv;

/*
 * prototypes
 */
int getln (char *, int);
static int chrcnv (char *);
static void chrdsp (char *, char);
static RETSIGTYPE intrser (int);


int
main (int argc, char **argv)
{
    int body = 1, prepend = 1, rapid = 0;
    int doteof = 0, fdi, fdo, i, state;
    char *cp, *drft = NULL, *erasep = NULL;
    char *killp = NULL, name[NAMESZ], field[BUFSIZ];
    char buffer[BUFSIZ], tmpfil[BUFSIZ];
    char **arguments, **argp;
    FILE *in, *out;

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* read user profile/context */
    context_read();

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++))
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    adios (NULL, "-%s unknown", cp);

		case HELPSW: 
		    snprintf (buffer, sizeof(buffer), "%s [switches] file",
			invo_name);
		    print_help (buffer, switches, 1);
		    done (1);
		case VERSIONSW:
		    print_version(invo_name);
		    done (1);

		case ERASESW: 
		    if (!(erasep = *argp++) || *erasep == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;
		case KILLSW: 
		    if (!(killp = *argp++) || *killp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;

		case PREPSW: 
		    prepend++;
		    continue;
		case NPREPSW: 
		    prepend = 0;
		    continue;

		case RAPDSW: 
		    rapid++;
		    continue;
		case NRAPDSW: 
		    rapid = 0;
		    continue;

		case BODYSW: 
		    body++;
		    continue;
		case NBODYSW: 
		    body = 0;
		    continue;

		case DOTSW: 
		    doteof++;
		    continue;
		case NDOTSW: 
		    doteof = 0;
		    continue;
	    }
	} else {
	    if (!drft)
		drft = cp;
	}

    if (!drft)
	adios (NULL, "usage: %s [switches] file", invo_name);
    if ((in = fopen (drft, "r")) == NULL)
	adios (drft, "unable to open");

    strncpy (tmpfil, m_tmpfil (invo_name), sizeof(tmpfil));
    if ((out = fopen (tmpfil, "w")) == NULL)
	adios (tmpfil, "unable to create");
    chmod (tmpfil, 0600);

    /*
     * Are we changing the kill or erase character?
     */
    if (killp || erasep) {
#ifdef HAVE_TERMIOS_H
	cc_t save_erase, save_kill;
#else
	int save_erase, save_kill;
#endif

	/* get the current terminal attributes */
#ifdef HAVE_TERMIOS_H
	tcgetattr(0, &tio);
#else
# ifdef HAVE_TERMIO_H
	ioctl(0, TCGETA, &tio);
# else
	ioctl (0, TIOCGETP, (char *) &tio);
	ioctl (0, TIOCGETC, (char *) &tc);
# endif
#endif

	/* save original kill, erase character for later */
	save_kill = KILL;
	save_erase = ERASE;

	/* set new kill, erase character in terminal structure */
	KILL = killp ? chrcnv (killp) : save_kill;
	ERASE = erasep ? chrcnv (erasep) : save_erase;

	/* set the new terminal attributes */
#ifdef HAVE_TERMIOS_H
	 tcsetattr(0, TCSADRAIN, &tio);
#else
# ifdef HAVE_TERMIO_H
	ioctl(0, TCSETAW, &tio);
# else
	ioctl (0, TIOCSETN, (char *) &tio);
# endif
#endif

	/* print out new kill erase characters */
	chrdsp ("erase", ERASE);
	chrdsp (", kill", KILL);
	chrdsp (", intr", INTR);
	putchar ('\n');
	fflush (stdout);

	/*
	 * We set the kill and erase character back to original
	 * setup in terminal structure so we can easily
	 * restore it upon exit.
	 */
	KILL = save_kill;
	ERASE = save_erase;
    }

    sigint = 0;
    SIGNAL2 (SIGINT, intrser);

    /*
     * Loop through the lines of the draft skeleton.
     */
    for (state = FLD;;) {
	switch (state = m_getfld (state, name, field, sizeof(field), in)) {
	    case FLD: 
	    case FLDEOF: 
	    case FLDPLUS: 
		/*
		 * Check if the value of field contains anything
		 * other than space or tab.
		 */
		for (cp = field; *cp; cp++)
		    if (*cp != ' ' && *cp != '\t')
			break;

		/* If so, just add header line to draft */
		if (*cp++ != '\n' || *cp != 0) {
		    printf ("%s:%s", name, field);
		    fprintf (out, "%s:%s", name, field);
		    while (state == FLDPLUS) {
			state =
			    m_getfld (state, name, field, sizeof(field), in);
			printf ("%s", field);
			fprintf (out, "%s", field);
		    }
		} else {
		    /* Else, get value of header field */
		    printf ("%s: ", name);
		    fflush (stdout);
		    i = getln (field, sizeof(field));
		    if (i == -1) {
abort:
			if (killp || erasep) {
#ifdef HAVE_TERMIOS_H
			    tcsetattr(0, TCSADRAIN, &tio);
#else
# ifdef HAVE_TERMIO
			    ioctl (0, TCSETA, &tio);
# else
			    ioctl (0, TIOCSETN, (char *) &tio);
# endif
#endif
			}
			unlink (tmpfil);
			done (1);
		    }
		    if (i != 0 || (field[0] != '\n' && field[0] != 0)) {
			fprintf (out, "%s:", name);
			do {
			    if (field[0] != ' ' && field[0] != '\t')
				putc (' ', out);
			    fprintf (out, "%s", field);
			} while (i == 1
				    && (i = getln (field, sizeof(field))) >= 0);
			if (i == -1)
			    goto abort;
		    }
		}

		if (state == FLDEOF) {	/* moby hack */
		    fprintf (out, "--------\n");
		    printf ("--------\n");
		    if (!body)
			break;
		    goto no_body;
		}
		continue;

	    case BODY: 
	    case BODYEOF:
	    case FILEEOF: 
	        if (!body)
	            break;
		fprintf (out, "--------\n");
		if (field[0] == 0 || !prepend)
		    printf ("--------\n");
		if (field[0]) {
		    if (prepend && body) {
			printf ("\n--------Enter initial text\n\n");
			fflush (stdout);
			for (;;) {
			    getln (buffer, sizeof(buffer));
			    if (doteof && buffer[0] == '.' && buffer[1] == '\n')
				break;
			    if (buffer[0] == 0)
				break;
			    fprintf (out, "%s", buffer);
			}
		    }

		    do {
			fprintf (out, "%s", field);
			if (!rapid && !sigint)
			    printf ("%s", field);
		    } while (state == BODY &&
			    (state = m_getfld (state, name, field, sizeof(field), in)));
		    if (prepend || !body)
			break;
		    else
			printf ("\n--------Enter additional text\n\n");
		}
no_body:
		fflush (stdout);
		for (;;) {
		    getln (field, sizeof(field));
		    if (doteof && field[0] == '.' && field[1] == '\n')
			break;
		    if (field[0] == 0)
			break;
 		    fprintf (out, "%s", field);
		}
		break;

	    default: 
		adios (NULL, "skeleton is poorly formatted");
	}
	break;
    }

    if (body)
	printf ("--------\n");

    fflush (stdout);
    fclose (in);
    fclose (out);
    SIGNAL (SIGINT, SIG_IGN);

    if (killp || erasep) {
#ifdef HAVE_TERMIOS_H
	 tcsetattr(0, TCSADRAIN, &tio);
#else
# ifdef HAVE_TERMIO_H
	ioctl (0, TCSETAW, &tio);
# else
	ioctl (0, TIOCSETN, (char *) &tio);
# endif
#endif
    }

    if ((fdi = open (tmpfil, O_RDONLY)) == NOTOK)
	adios (tmpfil, "unable to re-open");
    if ((fdo = creat (drft, m_gmprot ())) == NOTOK)
	adios (drft, "unable to write");
    cpydata (fdi, fdo, tmpfil, drft);
    close (fdi);
    close (fdo);
    unlink (tmpfil);

    context_save ();	/* save the context file */
    done (0);
}


int
getln (char *buffer, int n)
{
    int c;
    char *cp;

    cp = buffer;
    *cp = 0;

    switch (setjmp (sigenv)) {
	case OK: 
	    wtuser = 1;
	    break;

	case DONE: 
	    wtuser = 0;
	    return 0;

	default: 
	    wtuser = 0;
	    return NOTOK;
    }

    for (;;) {
	switch (c = getchar ()) {
	    case EOF: 
		clearerr (stdin);
		longjmp (sigenv, DONE);

	    case '\n': 
		if (cp[-1] == QUOTE) {
		    cp[-1] = c;
		    wtuser = 0;
		    return 1;
		}
		*cp++ = c;
		*cp = 0;
		wtuser = 0;
		return 0;

	    default: 
		if (cp < buffer + n)
		    *cp++ = c;
		*cp = 0;
	}
    }
}


static RETSIGTYPE
intrser (int i)
{
#ifndef	RELIABLE_SIGNALS
    SIGNAL (SIGINT, intrser);
#endif

    if (wtuser)
	longjmp (sigenv, NOTOK);
    sigint++;
}


static int
chrcnv (char *cp)
{
    return (*cp != QUOTE ? *cp : m_atoi (++cp));
}


static void
chrdsp (char *s, char c)
{
    printf ("%s ", s);
    if (c < ' ' || c == 0177)
	printf ("^%c", c ^ 0100);
    else
	printf ("%c", c);
}
